clang/runtime/dpct-rt/include/dpct/bindless_images.hpp

```diff
@@ -117,6 +151,349 @@ class image_mem_wrapper {
+  get_interop_mem_handle() const noexcept {
+    return *_interop_mem_handle_ptr;
+  }
+
+  sycl::ext::oneapi::experimental::image_mem_handle
+  get_img_mem_handle() const noexcept {
+    return _img_mem_wrapper_ptr->get_handle();
+  }
+
+  void *
+  get_buf_ptr() const noexcept {
+    return _buf_ptr;
+  }
+
+  sycl::ext::oneapi::experimental::image_descriptor
+  get_img_desc() const noexcept {
+    return _img_desc;
+  }
+
+  size_t get_res_size_bytes() {
+    return _res_size_bytes;
+  }
+
+  bool is_buffer_res() const {
+    return _res_is_buffer;
+  }
+
+  void map_resource(sycl::queue q) {
+    if (_res_is_buffer) {
+      if (_buf_ptr) {
+        throw std::runtime_error("Resource is already mapped!");
+      }
+
+      // NOTE: 'map_external_linear_memory' API is not yet merged by Codeplay
+      /*
+      _buf_ptr =
+          sycl::ext::oneapi::experimental::map_external_linear_memory(
+              get_interop_mem_handle(), _img_desc, q);
+      */
+      throw
+        std::runtime_error(
+            "Mapping a buffer resoure is not yet supported!");
+    }
+    else {
+      if (_img_mem_wrapper_ptr) {
+        throw std::runtime_error("Resource is already mapped!");
+      }
+
+      auto img_mem_handle =
+          sycl::ext::oneapi::experimental::map_external_image_memory(
+              get_interop_mem_handle(), _img_desc, q);
+      
+      _img_mem_wrapper_ptr =
+          new image_mem_wrapper(get_img_desc(), img_mem_handle);
+    }
+  }
+
+  image_mem_wrapper* get_mapped_mipmapped_array() {
+    if(_res_is_buffer) {
+      throw std::runtime_error(
+          "Buffer resouce cannot be accessed as an array!");
+    }
+    if(!_img_mem_wrapper_ptr) {
+      throw std::runtime_error("Resouce is not mapped!");
+    }
+
+    return _img_mem_wrapper_ptr;
+  }
+
+  image_mem_wrapper* get_sub_resource_mapped_array(
+        unsigned int array_index, unsigned int mip_level) {
+    if (_res_is_buffer) {
+      throw std::runtime_error(
+          "Buffer resouce cannot be accessed as an array!");
+    }
+    if (array_index != 0) {
+      throw std::runtime_error(
+          "Retrieving a particular layer from image array "
+          "is not yet supported!");
+    }
+    if (array_index >= _img_desc.array_size) {
+      throw std::runtime_error(
+          "Array index requested exceeds total array size in resource!");
+    }
+    if (mip_level >= _img_desc.num_levels) {
+        throw std::runtime_error(
+          "Mip level requested exceeds total num of levels in resource!");
+    }
+    if (!_img_mem_wrapper_ptr) {
+      throw std::runtime_error("Resouce is not mapped!");
+    }
+
+    return (mip_level == 0)?
+          _img_mem_wrapper_ptr: 
+          _img_mem_wrapper_ptr->get_mip_level(mip_level);
+  }
+
+  void unmap_resource(sycl::queue q) {
+    if (_res_is_buffer) {
+      if (!_buf_ptr) {
+        throw std::runtime_error("Resource is not mapped!");
+      }
+
+      _buf_ptr = nullptr;
```

Comment:
Should `_but_ptr` be freed before set null?

Reply:
The `_buf_ptr` obtained during mapping should not be freed because the underlying memory is created by DX11 and should be available to be used for DX11 APIs after unmapping of the resource CUDA/SYCL.